py/objtype.c

```diff
@@ -432,7 +432,7 @@ STATIC mp_obj_t instance_unary_op(mp_unary_op_t op, mp_obj_t self_in) {
                 // __hash__ must return a small int
                 val = MP_OBJ_NEW_SMALL_INT(mp_obj_get_int_truncated(val));
                 break;
-            case MP_UNARY_OP_INT:
+            case MP_UNARY_OP_INT_MAYBE:
                 // Must return int
                 if (!mp_obj_is_int(val)) {
                     mp_raise_TypeError(NULL);
```

Comment:
Should this be changed to `return MP_OBJ_NULL` now instead of raising to match the semantics of "maybe"?

Reply:
If we get here then the user class defined `__int__` so it does support the method.  But it returned a non-integer type which is illegal.  Hence the raising of an error.